stingray/covariancespectrum.py

```diff
@@ -0,0 +1,428 @@
+
+        energy_covar : dictionary
+            A dictionary with mid point of band_interest and their covariance
+            computed with their individual reference band. The covaraince
+            values are normalized.
+
+        unnorm_covar : np.ndarray
+            An array of arrays with mid point band_interest and their
+            covariance. It is the array-form of the dictionary `energy_covar`.
+            The covariance values are unnormalized.
+
+        covar : np.ndarray
+            Normalized covaraiance spectrum.
+
+        covar_error : np.ndarray
+            Errors of the normalized covariance spectrum.
+
+        min_time : int
+            Time of arrival of the earliest photon.
+
+        max_time : int
+            Time of arrival of the last photon.
+
+        min_energy : float
+            Energy of the photon with the minimum energy.
+
+        max_energy : float
+            Energy of the photon with the maximum energy.
+
+        Reference
+        ---------
+        [1] Wilkinson, T. and Uttley, P. (2009), Accretion disc variability
+            in the hard state of black hole X-ray binaries. Monthly Notices
+            of the Royal Astronomical Society, 397: 666–676.
+            doi: 10.1111/j.1365-2966.2009.15008.x
+        """
+
+        self.avg_covar = False  # This parameter is used for the purpose of
+                                # AveragedCovariancespectrum.
+
+        self._init_vars(event_list, dt, band_interest,
+                        ref_band_interest, std)
+
+        # A dictionary with energy bin as key and events as value of the key
+        self.energy_events = {}
+
+        self._construct_energy_events(self.energy_events)
+
+        self._update_energy_events(self.energy_events)
+
+        # The dictionary with covariance spectrum for each energy bin
+        self.energy_covar = {}
+
+        self._construct_energy_covar(self.energy_events, self.energy_covar)
+
+    def _init_vars(self, event_list, dt, band_interest,
+                   ref_band_interest, std):
+        if np.all(np.diff(event_list, axis=0).T[0] >= 0) == False:
+            utils.simon("The event list must be sorted with respect to times of arrivals.")
+            event_list = event_list[event_list[:, 0].argsort()]
+
+        self.event_list = event_list
+
+        self.event_list_T = event_list.T
+
+        self._init_special_vars()
+
+        if ref_band_interest is None:
+            ref_band_interest = (self.min_energy, self.max_energy)
+
+        assert type(ref_band_interest) in (list, tuple), "Ref Band interest " \
+                                                         "should be either " \
+                                                         "tuple or list."
+
+        assert len(ref_band_interest) == 2, "Band interest should be a tuple" \
+                                            " with min and max energy value " \
+                                            "for the reference band."
```

Comment:
Technically, it will raise the "AssertionError" but I thought we'll be refactoring the entire Stingray codebase while solving #112.